FIX Check and correct the input_tags.sparse flag

antoinebaker opened the pull request:
#### Reference Issues/PRs

Fixes #30139


#### What does this implement/fix? Explain your changes.

The test `check_estimator_sparse_tag` is added.
It checks that the `input_tags.sparse` tag of an estimator is consistent:
- if  `input_tags.sparse=True` the estimator can be fitted on sparse data
- if `input_tags.sparse=False` the estimator must raise a error when fitted on sparse data

The `input_tags.sparse` flag was edited for a lot of estimators.

antoinebaker committed:
add input_tags.sparse and test

antoinebaker committed:
fix LinearRegression tag

antoinebaker committed:
changelog

antoinebaker committed:
Merge branch 'main' into fix_input_sparse_tag

ogrisel commented on doc/whats_new/upcoming_changes/sklearn.utils/30187.api.rst:
```diff
@@ -0,0 +1,4 @@
+- :func:`utils.estimator_checks.check_estimator_sparse_tag` ensures that
+  the estimator tag `input_tags.sparse` is consistent with its `fit`
+  method (acceting sparse input `X` or raising an error).
+  By :user:`Antoine Baker <antoinebaker>`
```
Let's include the list of all estimators whose `tags.input_tags.sparse` value was fixed here.

Maybe we could introduce a new subdirectory under `doc/whats_new/upcoming_changes/` dedicated to PRs impacting many modules at once.

WDYT @lesteve @adrinjalali?

ogrisel replied:
There is already such a subdirectory named, for instance this particular PR should probably have an entry in `changed-models` to document the fix to all impacted estimators (with the big list) and maybe a second "enhancement" entry under `sklearn.utils` to document the enhancement to the estimator check function.

antoinebaker replied:
Is that okay that the list will be real big (~100 estimators) ?

ogrisel replied:
Indeed if the list that big, it's fine to state that this changes impacts many estimators (even "a majority of estimators" in scikit-learn?)

ogrisel reviewed (commented):
Thanks for the PR. Any reason to decide to introduce a new `check_estimator_sparse_tag`estimator check instead of making the existing `check_estimator_sparse_array` tag-aware?

Since estimator checks are already costly to run, I would rather avoid proliferation and redundancy when possible.

ogrisel commented on doc/whats_new/upcoming_changes/sklearn.utils/30187.api.rst:
```diff
@@ -0,0 +1,4 @@
+- :func:`utils.estimator_checks.check_estimator_sparse_tag` ensures that
```
The filename is `30187.api.rst` but I think it should be considered a fix since those tag values were wrong/missing.

ogrisel replied:
This comment is now obsolete. Please refer to https://github.com/scikit-learn/scikit-learn/pull/30187#discussion_r1827900270 instead.

antoinebaker committed:
changelog

antoinebaker commented:
> Thanks for the PR. Any reason to decide to introduce a new `check_estimator_sparse_tag`estimator check instead of making the existing `check_estimator_sparse_array` tag-aware?
> 
> Since estimator checks are already costly to run, I would rather avoid proliferation and redundancy when possible.

I tried to make `check_estimator_sparse_array` tag-aware, but didn't manage to combine the old behavior (testing on various sparse formats while optionally ignoring an error if the estimator does not accept this sparse format) and what the new test is doing (checking that `sparse` tag is equivalent to raising or not the error).

antoinebaker committed:
fix column transformer tag

antoinebaker committed:
change error message

antoinebaker committed:
changelog

antoinebaker committed:
Merge remote-tracking branch 'upstream/main' into fix_input_sparse_tag

glemaitre commented on doc/whats_new/upcoming_changes/sklearn.utils/30187.enhancement.rst:
```diff
@@ -0,0 +1,4 @@
+- :func:`utils.estimator_checks.check_estimator_sparse_tag` ensures that
+  the estimator tag `input_tags.sparse` is consistent with its `fit`
+  method (acceting sparse input `X` or raising the appropriate error).
```
```suggestion
  method (accepting sparse input `X` or raising the appropriate error).
```

glemaitre commented on doc/whats_new/upcoming_changes/changed-models/30187.fix.rst:
```diff
@@ -0,0 +1,3 @@
+- The `tags.input_tags.sparse` flag was corrected for a
+  majority of estimators.
```
```suggestion
- The `tags.input_tags.sparse` flag was corrected for a majority of estimators.
```

glemaitre commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1192,6 +1193,39 @@ def check_array_api_input_and_values(
     )
 
 
+def check_estimator_sparse_tag(name, estimator_orig):
```
```suggestion
def check_estimator_sparse_tag(name, estimator_orig):
    """Check that estimator tag related with accepting sparse data is properly set."""
```

glemaitre commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1192,6 +1193,39 @@ def check_array_api_input_and_values(
+
+    tags = get_tags(estimator)
+    if tags.input_tags.sparse:
+        estimator.fit(X, y)  # should pass
```
I'm thinking that we could help third-party library a bit better here by re-raising a better error message.

```suggestion
        try:
            estimator.fit(X, y)  # should pass
        except Exception as e:
            raise AssertionError(
                f"Estimator {name} raised an exception: {e}. The tag "
                "self.input_tags.sparse_tag might not be consistent with the "
                "estimator's ability to handle sparse data (i.e. controlled by the "
                "parameter `accept_sparse` in `validate_data` or `check_array` "
                f"functions). Got input_tags.sparse={tags.input_tags.sparse}."
            )
```

glemaitre commented on sklearn/utils/tests/test_estimator_checks.py:
I think that we could add a test for the check as well:

```python
def test_check_estimator_sparse_tag():
    """Test that check_estimator_sparse_tag raises error when sparse tag is
    misaligned."""

    class EstimatorWithSparseConfig(BaseEstimator):
        def __init__(self, tag_sparse, accept_sparse):
            self.tag_sparse = tag_sparse
            self.accept_sparse = accept_sparse

        def fit(self, X, y=None):
            validate_data(self, X, y, accept_sparse=self.accept_sparse)
            return self

        def __sklearn_tags__(self):
            tags = super().__sklearn_tags__()
            tags.input_tags.sparse = self.tag_sparse
            return tags

    test_cases = [
        {"tag_sparse": True, "accept_sparse": True, "error_type": None},
        {"tag_sparse": False, "accept_sparse": False, "error_type": None},
        {"tag_sparse": False, "accept_sparse": True, "error_type": AssertionError},
        {"tag_sparse": True, "accept_sparse": False, "error_type": AssertionError},
    ]

    for test_case in test_cases:
        estimator = EstimatorWithSparseConfig(
            test_case["tag_sparse"], test_case["accept_sparse"]
        )
        if test_case["error_type"] is None:
            check_estimator_sparse_tag(estimator.__class__.__name__, estimator)
        else:
            with raises(test_case["error_type"]):
                check_estimator_sparse_tag(estimator.__class__.__name__, estimator)
```

This might looks like an overkill to test the check that should check the estimator ;P but this is sometimes useful for the future us.

antoinebaker replied:
done in d6f277

glemaitre reviewed (commented):
This a first round of review just looking at the test. I'll check now the changes in the different estimator to know if I'm getting surprise by what I'll see.

glemaitre commented on sklearn/decomposition/_incremental_pca.py:
```diff
@@ -418,3 +418,8 @@ def transform(self, X):
+
+    def __sklearn_tags__(self):
+        tags = super().__sklearn_tags__()
+        tags.input_tags.sparse = True
```
This one is really funky: `fit` accepts sparse but not `partial_fit`. So I assume that we can be lenient because we cannot use `partial_fit` in meta-estimator or pipeline.

glemaitre replied:
This could be good to add a small comment to acknowledge this specificity.

antoinebaker replied:
done in 705c414

glemaitre commented on sklearn/decomposition/_pca.py:
```diff
@@ -851,4 +851,5 @@ def __sklearn_tags__(self):
         tags = super().__sklearn_tags__()
         tags.transformer_tags.preserves_dtype = ["float64", "float32"]
         tags.array_api_support = True
+        tags.input_tags.sparse = True
```
Here, it will depends on the solver:


```suggestion
        tags.input_tags.sparse = self.svd_solver in ("auto", "arpack", "covariance_eigh")
```

I just realise that `score_samples` does not accept at all sparse data. We might check if it is still the case with the improvement recently done with the `"covariance_eigh"` solver or if we can go around the original issue.

glemaitre commented on sklearn/ensemble/_gb.py:
```diff
@@ -1725,6 +1725,11 @@ def staged_predict_proba(self, X):
                 "loss=%r does not support predict_proba" % self.loss
             ) from e
 
+    def __sklearn_tags__(self):
```
Isn't it possible to have a single `__sklearn_tags__(self)` defined in `BaseGradientBoosting`?

antoinebaker replied:
done in 705c414

glemaitre commented on sklearn/ensemble/_weight_boosting.py:
```diff
@@ -1166,3 +1171,8 @@ def staged_predict(self, X):
         for i, _ in enumerate(self.estimators_, 1):
             yield self._get_median_predict(X, limit=i)
+
+    def __sklearn_tags__(self):
```
Same here, shall we have a single shared definition in `BaseWeightBoosting`?

antoinebaker replied:
done in 705c414

glemaitre commented on sklearn/linear_model/_base.py:
```diff
@@ -687,6 +687,11 @@ def rmatvec(b):
         self._set_intercept(X_offset, y_offset, X_scale)
         return self
 
+    def __sklearn_tags__(self):
```
I think that those could be moved in the `LinearModelCV` because `fit` and the `validate_data` is defined there.

glemaitre replied:
It should be done for all models inheriting from `LinearModelCV`.

antoinebaker replied:
But `LinearRegression` inherits from `LinearModel` and not `LinearModelCV` ?

glemaitre replied:
Oups, I might have commented on the wrong line. I recall to have seen it for `ElasticNetCV` and `LassoCV` if I recall properly.

glemaitre commented on sklearn/linear_model/_coordinate_descent.py:
```diff
@@ -2078,6 +2083,7 @@ def _is_multitask(self):
 
     def __sklearn_tags__(self):
         tags = super().__sklearn_tags__()
+        tags.input_tags.sparse = True
```
As mentioned somewhere else, it make smore sense to defined next to the `fit` so in `LinearModelCV`.

antoinebaker replied:
done in 705c414

glemaitre commented on sklearn/linear_model/_coordinate_descent.py:
```diff
@@ -2359,6 +2365,7 @@ def _is_multitask(self):
 
     def __sklearn_tags__(self):
         tags = super().__sklearn_tags__()
+        tags.input_tags.sparse = True
```
Same here. Move it to `LinearModelCV`.

antoinebaker replied:
done in 705c414

glemaitre commented on sklearn/linear_model/_coordinate_descent.py:
```diff
@@ -2078,6 +2083,7 @@ def _is_multitask(self):
     def __sklearn_tags__(self):
         tags = super().__sklearn_tags__()
+        tags.input_tags.sparse = True
         tags.target_tags.multi_output = False
```
I'm even wondering if the multi_output could also be moved together with teh `_is_multitask` method but it needs to be checked.

antoinebaker replied:
done in 3ac38e1

glemaitre commented on sklearn/linear_model/_ransac.py:
```diff
@@ -721,3 +721,8 @@ def get_metadata_routing(self):
+
+    def __sklearn_tags__(self):
+        tags = super().__sklearn_tags__()
+        tags.input_tags.sparse = True
```
I agree that `fit` involved this change. However, I'm wondering if we should not instead delegate to the underlying estimator instead: if the underlying estimator does not support it, this `RANSACRegressor` will fail (e.g. `LinearRegression(positive_only=True)`).

antoinebaker replied:
done in 8f2f3db

glemaitre commented on sklearn/naive_bayes.py:
```diff
@@ -880,6 +880,7 @@ def __init__(
 
     def __sklearn_tags__(self):
         tags = super().__sklearn_tags__()
+        tags.input_tags.sparse = True
```
It seems that this is shared with `_BaseDiscreteNB`. So we can define it once only in the base class since the validation of `fit` is located there.

antoinebaker replied:
done in 705c414

glemaitre commented on sklearn/pipeline.py:
```diff
@@ -1224,6 +1224,7 @@ def __sklearn_tags__(self):
                 tags.input_tags.pairwise = get_tags(
                     self.steps[0][1]
                 ).input_tags.pairwise
+                tags.input_tags.sparse = get_tags(self.steps[0][1]).input_tags.sparse
```
Ouch I don't think this is that simple. To be complete, we should check the sparse of each step and somewhow the output as well. If the previous output agree with the next input then we would be fine.

Somehow, we might want to be lenient and set it to True and hope that the underlying estimator will just raise.

I would need a second opinion here.

antoinebaker replied:
Maybe @adrinjalali ?

glemaitre replied:
Yep indeed I would do something similar to what you did for the `FeatureUnion` and checking that each steps accept sparse data.

antoinebaker replied:
done in a9fb7d7

glemaitre commented on sklearn/pipeline.py:
I think that we are missing `FeatureUnion` here.

glemaitre commented on sklearn/preprocessing/_data.py:
`RobustScaler` should accept sparse data but it is not set for the moment.

antoinebaker replied:
done in e85f94a

glemaitre commented on sklearn/preprocessing/_function_transformer.py:
```diff
@@ -394,6 +394,7 @@ def __sklearn_tags__(self):
         tags = super().__sklearn_tags__()
         tags.no_validation = not self.validate
         tags.requires_fit = False
+        tags.input_tags.sparse = True
```
This should be a bit more tricky because it will depend on `self.validate` and `self.accept_sparse`


```suggestion
        tags.input_tags.sparse = not self.validate or (self.validate and self.accept_sparse)
```

glemaitre commented on sklearn/semi_supervised/_self_training.py:
```diff
@@ -7,7 +7,7 @@
 
 from ..base import BaseEstimator, MetaEstimatorMixin, _fit_context, clone
-from ..utils import Bunch, safe_mask
+from ..utils import Bunch, get_tags, safe_mask
```
Oh, could you modify the import above and import `ClassifierMixin` from the relative import `..base`.

glemaitre commented on sklearn/svm/_classes.py:
```diff
@@ -1140,6 +1155,11 @@ def __init__(
             random_state=random_state,
         )
 
+    def __sklearn_tags__(self):
```
This could be set in `BaseSVC` instead.

antoinebaker replied:
done in 705c414

glemaitre commented on sklearn/svm/_classes.py:
```diff
@@ -877,6 +887,11 @@ def __init__(
             random_state=random_state,
         )
 
+    def __sklearn_tags__(self):
```
This could be set in `BaseSVC` instead.

antoinebaker replied:
done in 705c414

glemaitre commented on sklearn/svm/_classes.py:
```diff
@@ -1334,6 +1354,11 @@ def __init__(
         )
 
+    def __sklearn_tags__(self):
+        tags = super().__sklearn_tags__()
```
This could be set in `BaseLibSVM` if possible.

glemaitre commented on sklearn/svm/_classes.py:
```diff
@@ -1521,6 +1546,11 @@ def __init__(
             random_state=None,
         )
 
+    def __sklearn_tags__(self):
```
This could be set in BaseLibSVM if possible.

glemaitre commented on sklearn/svm/_classes.py:
```diff
@@ -1777,3 +1807,8 @@ def predict(self, X):
         y = super().predict(X)
         return np.asarray(y, dtype=np.intp)
+
+    def __sklearn_tags__(self):
```
This could be set in BaseLibSVM if possible.

antoinebaker replied:
done in 705c414

glemaitre commented on sklearn/tree/_classes.py:
```diff
@@ -1100,6 +1100,7 @@ def __sklearn_tags__(self):
         }
         tags.classifier_tags.multi_label = True
         tags.input_tags.allow_nan = allow_nan
+        tags.input_tags.sparse = True
```
This could be set in the in the base class as well and for the decision tree regressor.

antoinebaker replied:
done in 705c414

glemaitre reviewed (commented):
OK I think that I made a pass on all estimators. I think that sometimes, we can move the definition in the base class.

antoinebaker committed:
fix passthrough sparse tag

antoinebaker committed:
fix SelfTrainingClassifier

antoinebaker committed:
Apply suggestions from code review

Co-authored-by: Guillaume Lemaitre <guillaume@probabl.ai>

antoinebaker committed:
add suggestions

antoinebaker committed:
fix multitask

antoinebaker committed:
black formatting

antoinebaker committed:
Merge branch 'main' into fix_input_sparse_tag

glemaitre commented:
@antoinebaker do not hesitate to ping me when you want another review.

antoinebaker committed:
add meta test

antoinebaker committed:
add feature union

antoinebaker committed:
check function transformer

antoinebaker commented:
Hello @glemaitre thanks for the review ! I think I addressed all the suggestions. You might want to take a look at `Pipeline`,  `FeatureUnion`, `ColumnTransformer` and `_BaseHeterogeneousEnsemble`, I have some doubts regarding these.

antoinebaker commented on sklearn/pipeline.py:
```diff
@@ -2113,6 +2114,15 @@ def get_metadata_routing(self):
 
         return router
 
+    def __sklearn_tags__(self):
+        tags = super().__sklearn_tags__()
+        tags.input_tags.sparse = all(
+            get_tags(trans).input_tags.sparse
+            for name, trans in self.transformer_list
+            if trans not in {"passthrough", "drop"}
+        )
+        return tags
+
```
@glemaitre should I be defensive against invalid `transformers_list` with a `try/except` as is done for `Pipeline` and `_BaseHeterogeneousEnsemble` ?

antoinebaker replied:
The same holds for `ColumnTransformer` and `self.transformers`.

glemaitre replied:
Yep because you can end-up in this function before to have call the `fit` function that is in charge of validating the `self.transformer_list`.

antoinebaker replied:
done in 21ec585

glemaitre commented:
> think I addressed all the suggestions. You might want to take a look at Pipeline, FeatureUnion, ColumnTransformer and _BaseHeterogeneousEnsemble, I have some doubts regarding these.

I'll have another look. I'm currently dealing with similar issue I think so once done on my side, it might be clearer what to do :)

antoinebaker committed:
catch invalid transformers list

glemaitre committed:
Merge branch 'main' into fix_input_sparse_tag

glemaitre commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1196,6 +1197,64 @@ def check_array_api_input_and_values(
+                f"functions). Got input_tags.sparse={tags.input_tags.sparse}."
+            )
+    else:
+        try:
+            estimator.fit(X, y)  # should fail with appropriate error
+        except Exception as e:
+            if _is_sparse_input_error(e):
+                return
+            else:
+                raise AssertionError(
+                    f"Estimator {name} raised an exception: {e}. "
+                    "The estimator failed when fitted on sparse data in accordance "
+                    f"with its tag self.input_tags.sparse={tags.input_tags.sparse} "
+                    "but didn't raise the appropriate error : error message should "
+                    "state explicitly that sparse input is not supported if this is "
+                    "not the case, e.g. by using check_array(X, accept_sparse=False)."
+                )
```
I would find it easier to read without the call to the outside function

```suggestion
        err_msg = (
            f"Estimator {name} raised an exception: {e}. "
            "The estimator failed when fitted on sparse data in accordance "
            f"with its tag self.input_tags.sparse={tags.input_tags.sparse} "
            "but didn't raise the appropriate error : error message should "
            "state explicitly that sparse input is not supported if this is "
            "not the case, e.g. by using check_array(X, accept_sparse=False)."
        )
        try:
            estimator.fit(X, y)  # should fail with appropriate error
        except (ValueError, TypeError) as e:
            if re.search("[Ss]parse", str(e)):
                # Got the right error type and mentioning sparse issue
                return
            raise AssertionError(err_msg) from e
        except:
            raise AssertionError(err_msg) from e
```

antoinebaker replied:
done in d9f4de3

glemaitre commented on sklearn/semi_supervised/_self_training.py:
```diff
@@ -613,3 +617,9 @@ def get_metadata_routing(self):
+
+    def __sklearn_tags__(self):
+        tags = super().__sklearn_tags__()
+        if self.estimator is not None:
```
We might want to add a `# TODO(1.8): remove together with base_estimator`.
Basically, we will remove the `None` option for `estimator` so the `if` will not be necessary.

antoinebaker replied:
done in c653834

glemaitre reviewed (commented):
It looks good. I'm only wondering about the `Pipeline` where I would do the check the different steps. It seems more logical to me.

antoinebaker committed:
add todo

antoinebaker committed:
remove outer function

antoinebaker committed:
change pipeline tag

antoinebaker commented:
I followed @glemaitre suggestion for setting the `sparse` tag in `Pipeline`. 

We now define`sparse` = all steps accepts sparse instead of the previous one `sparse` = first non passthrough step accepts sparse. Both are incorrect, but determining the correct flag seems a nightmare in general as some transformers may change the sparse nature of the data. 

As a lesser of two evils the new definition will have false negatives (Pipelines accepting sparse data wrongly tagged `sparse=False`) while the old one have false positives (Pipelines rejecting sparse data wrongly tagged `sparse=True`).

antoinebaker committed:
tag RobustScaler

antoinebaker committed:
tag RANSAC

antoinebaker committed:
multi_output in LinearModelCV

antoinebaker commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1196,6 +1197,61 @@ def check_array_api_input_and_values(
+        try:
+            estimator.fit(X, y)  # should pass
+        except Exception as e:
+            raise AssertionError(
+                f"Estimator {name} raised an exception: {e}. The tag "
+                "self.input_tags.sparse might not be consistent with the "
+                "estimator's ability to handle sparse data (i.e. controlled by the "
+                "parameter `accept_sparse` in `validate_data` or `check_array` "
+                f"functions). Got input_tags.sparse={tags.input_tags.sparse}."
+            )
```
Should I maybe `raise AssertionError(err_msg) from e` as we did for the `tags.input_tags.sparse=False` case ?

glemaitre replied:
Yes I like this suggestion to have a proper traceback. This is a developer traceback so we can be verbose at this stage.

antoinebaker replied:
done in 425b473

glemaitre reviewed (approved):
There a small conflict to solve.

Otherwise, it looks good to me. I'm fine with the current policy for the `Pipeline` here.

antoinebaker committed:
Merge remote-tracking branch 'upstream/main' into fix_input_sparse_tag

antoinebaker committed:
raise from exception

antoinebaker committed:
no cover

antoinebaker committed:
test raise inappropriate error

antoinebaker committed:
Merge branch 'main' into fix_input_sparse_tag

jeremiedbb commented on sklearn/pipeline.py:
```diff
@@ -1226,6 +1226,15 @@ def __sklearn_tags__(self):
                 tags.input_tags.pairwise = get_tags(
                     self.steps[0][1]
                 ).input_tags.pairwise
+            # WARNING: the sparse tag can be incorrect.
+            # Some Pipelines accepting sparse data are wrongly tagged sparse=False.
+            # For example Pipeline([PCA(), estimator]) accepts sparse data
+            # even if the estimator doesn't as PCA outputs a dense array.
+            tags.input_tags.sparse = all(
+                get_tags(step).input_tags.sparse
+                for name, step in self.steps
+                if step != "passthrough"
+            )
```
This makes me wonder if we aren't expecting too much from the sparse tag. Maybe we should narrow its scope to the estimator level, regardless of whether or not its sub-estimator(s), if any, do support sparse input. In that case sparse means "this estimator can handle sparse input but its inner estimator(s) (if it's a meta-estimator) might not in which case it's undefined behavior". What do you think @glemaitre and @adrinjalali ?

jeremiedbb replied:
As discussed irl with Guillaume, an other option is to change `sparse` to accept ``True``, ``False`` and ``Maybe`` (or ``Undefined``, name tbd).

jeremiedbb commented on sklearn/pipeline.py:
```diff
@@ -2115,6 +2124,21 @@ def get_metadata_routing(self):
 
         return router
 
+    def __sklearn_tags__(self):
+        tags = super().__sklearn_tags__()
+        try:
+            tags.input_tags.sparse = all(
+                get_tags(trans).input_tags.sparse
+                for name, trans in self.transformer_list
+                if trans not in {"passthrough", "drop"}
+            )
+        except Exception:
+            # If `transformer_list` does not comply with our API (list of tuples)
+            # then it will fail. In this case, we assume that `sparse` is False
+            # but the parameter validation will raise an error during `fit`.
+            pass  # pragma: no cover
+        return tags
```
my previous comment is also motivated by this pattern that shows-up a few times. We're trying to inspect the full compute graph ahead of time which is something scikit-learn is really not good at by design.

jeremiedbb commented on sklearn/semi_supervised/_self_training.py:
```diff
@@ -613,3 +617,10 @@ def get_metadata_routing(self):
+
+    def __sklearn_tags__(self):
+        tags = super().__sklearn_tags__()
+        # TODO(1.8): remove together with base_estimator
```
it wasn't clear to what to remove in 1.8 until I read Guillaume's comment.
```suggestion
        # TODO(1.8): remove the condition check together with base_estimator
```

jeremiedbb commented:
Also, I think we should include it in 1.6.1 because it fixes tags that we just made public

antoinebaker committed:
Apply suggestions from code review

Co-authored-by: Jérémie du Boisberranger <jeremie@probabl.ai>

antoinebaker committed:
suggestions from code review

antoinebaker commented on sklearn/linear_model/_ridge.py:
```diff
@@ -1568,6 +1572,14 @@ def fit(self, X, y, sample_weight=None):
+        reject_sparse = (self.solver == "svd") or (
+            self.solver == "cholesky" and self.fit_intercept
+        )
+        tags.input_tags.sparse = not reject_sparse
```
done in 66c3bd9

antoinebaker commented on sklearn/ensemble/_forest.py:
```diff
@@ -1001,7 +1001,7 @@ def predict_log_proba(self, X):
     def __sklearn_tags__(self):
         tags = super().__sklearn_tags__()
-        tags.classifier_tags.multi_label = True
+        tags.input_tags.sparse = True
```
done in 66c3bd9

jeremiedbb committed:
Merge remote-tracking branch 'upstream/main' into pr/antoinebaker/30187

jeremiedbb reviewed (approved):
LGTM.

I think there are still things to improve with the sparse tag but it can be done in a separate PR. I'm in favor of merging this one as is to not delay the 1.6.1 release further.

glemaitre commented:
> I think there are still things to improve with the sparse tag but it can be done in a separate PR. I'm in favor of merging this one as is to not delay the 1.6.1 release further.

Yep it is indeed better than what we had. We can later improve and discussed having a ternary option instead of the binary one.

ItsIronOxide commented on sklearn/linear_model/_base.py:
```diff
@@ -689,7 +689,7 @@ def rmatvec(b):
     def __sklearn_tags__(self):
         tags = super().__sklearn_tags__()
-        tags.input_tags.sparse = True
+        tags.input_tags.sparse = not self.positive
```
My team changed to scikit-learn v1.6.1 this week. We had v1.5.1 before. Our code crashes in this exact line with the error "Unexpected <class 'AttributeError'>. 'LinearRegression' object has no attribute 'positive'".

We cannot deploy in production because of this. I am desperate enough to come here to ask for help. I do not understand why it would complain that the attribute does not exist given that we were using v1.5.1 before and the attribute has existed for 4 years now.

adrinjalali replied:
@ItsIronOxide feel free to open a new issue with a reproducer, and ping me. Happy to have a look and help out.

ItsIronOxide reviewed (commented):
Any help or guidance on this issue is very much appreciated.
